### PR TITLE
Add chemical/magical fires

### PR DESCRIPTION
**Pull request by FlameArrow57**
[GAME OBJECTS][FEATURE]

## About the PR 
This PR adds a new version of fire, chemical fire. Chemical fires are a resprited fire type intended to represent a fire created from a chemical or magical source, like wizard fireballs, that comes off the ground.

-They act the same as atmospheric fire and can be on the same tile. 
-They have two layered parts to them, one that appears below objects on their turf, and one that appears over.
-Chemical fire color can be something other than red, dependent on the source. Ex. magnesium fires will burn white, potassium fires will burn purple.

This PR replaces all non-atmospheric fires with it

Also makes it so that the Wizard fireball spell will now leave fire behind that is the color of the Wizard's hat.

Note-
I realize that this could tie in more closely with chemicals. It was suggested on Discord these fires could relate to chemical reagents much more closely, such as duration dependent on chemical volume on the source turf, reactants available and their types (oxidizer, etc), and resultant fire temperature in relation to chemicals. However, I think that would really be a larger rework/rebalance that this PR wasn't really aimed at initially. This PR was mainly to focus on the visual part of fires created through /proc/fireflash().

image

image

image

image

image

image

image

## Why's this needed? 
All fires in the game are represented as atmospheric fires when some are really a chemical fire, so more accurate in that sense.

Adds a new 3/4s perspective version of fire.

Allows for colored fire

## Changelog 
```changelog
(u)FlameArrow57, Flaborized, and Azrun
(*)Added a new type of fire, chemical fire, with sprites by Flaborized! This is a type of fire that acts the same as atmospheric fire but comes from chemical/magical sources, colored depending on the source, and can be on the same tile as atmospheric fire. Replaces all non-atmospheric fires.
(+)Fire created from the Wizard fireball spell will now be the color of the hat the Wizard is wearing.
```

### CODE REVIEW

**Comment by FlameArrow57** [comment]
PR'd for the moment just to get thoughts/approval on the idea. Still some code to test more and needs improvement on sprites.

**Comment by FlameArrow57** [comment]
Sure

**Comment by PrimeNumb** [comment]
Could you add screenshots of the new assets?

**Comment by FlameArrow57** [comment]
Sure, they're just placeholders for the moment though

**Comment by colossusqw** [comment]
Wait, does that mean we can get burning oil spills and horrible napalm leakages now? If so, fantastic.

**Comment by FlameArrow57** [comment]
Yup

File: code/modules/status_system/statusEffects.dm
Comment (Tarmunora) [inline code/modules/status_system/statusEffects.dm:2465]: this is... a bit cursed

File: code/obj/chem_fire.dm
Comment (Tarmunora) [inline code/obj/chem_fire.dm:44]: since you're running processing on these, should track remaining duration and kill in the process instead of using a SPAWN here

File: code/modules/antagonists/wizard/abilities/phaseshift.dm
Comment (Tarmunora) [inline code/modules/antagonists/wizard/abilities/phaseshift.dm:366]: should always use the chemfireflash proc instead of directly instantiating the fires

**Review by Tarmunora — CHANGES_REQUESTED** [review]
Not sure how this is going to play out - feels weird to have a second separate implementation of fire 🤷

File: code/modules/status_system/statusEffects.dm
Comment (FlameArrow57) [reply code/modules/status_system/statusEffects.dm:2465]: /datum/controller/process instead maybe?

**Comment by CheffieGithub** [comment]
I like them for the visuals since hotspots look strange not as part of atmos

File: code/modules/status_system/statusEffects.dm
Comment (Tarmunora) [reply code/modules/status_system/statusEffects.dm]: 🤷

**Comment by FlameArrow57** [comment]
> Not sure how this is going to play out - feels weird to have a second separate implementation of fire 🤷

Still WIP, but ended up reworking it reuse type /obj/hotspot still, with allowance for chemical and atmospheric files on the same tile

**Comment by FlameArrow57** [comment]
Okay I updated the PR description to the full intent of the PR, adding notes about scope. Code is finished, though the lighting effects might be too strong. Only thing not finished is sprites, which will be done pending approval of the idea

**Comment by Valtsu0** [comment]
A colorable sprite would be really nice, but it would probably be quite hard to make it look good

**Comment by FlameArrow57** [comment]
> A colorable sprite would be really cool, but it would probably be quite hard to make it look good

If you mean a single sprite that can be recolored through code instead of separate color sprites, I would agree, but I think it's a bit hard to do that/not possible since recoloring the whole sprite of the fire doesn't capture the different shades well (ex. red, orange, yellow for red fire)

**Comment by FlameArrow57** [comment]
Okay I'm marking it ready for review, only thing that's not finished is the sprites. Flaborized said they'd be interested pending approval of the code/idea

**Review by Tarmunora — CHANGES_REQUESTED** [review]

**Comment by FlameArrow57** [comment]
Okay new sprites are added in and code is finished, ready for re-review whenever

**Comment by NightmareChamillian** [comment]
Would it be possible to have a grayscale, colorable sprite included for future compatibility?

**Comment by FlameArrow57** [comment]
> Would it be possible to have a grayscale, colorable sprite included for future compatibility?

Sort of, it'd be trivial to add I guess, I just think it wouldn't capture different color shades very well

**Comment by FlameArrow57** [comment]
Pinged Yass in Discord and it sounded like this would "probably a good change". No feedback since then. Closing since I don't want to maintain this anymore, but mergeable/working as-is, and would be trivial to add some transparency or blending to the fire if needed

**Comment by FlameArrow57** [comment]
@Tarmunora Hey, does this sound interesting still by any chance? It's pretty much finished. If it still sounds good, was thinking some transparency could be added to the sprites or something, otherwise wasn't sure if the sprites were too big or if there were any problems

**Comment by Tarmunora** [comment]
Could be worth a TM at least, yeah

**Comment by TDHooligan** [comment]
is it feasible to add some overlay or sprite change remove the tiling? that's my one holdup here versus stuff like hotspots.

**Comment by FlameArrow57** [comment]
> Could be worth a TM at least, yeah

Kay, re-opened and it works with merge from master. LMK if you want any changes!

**Comment by FlameArrow57** [comment]
> is it feasible to add some overlay or sprite change to hide/obfuscate the really visible tiling? that's my one holdup here versus stuff like hotspots.

Maybe, I think it would require Flaborized redoing the sprites a little bit or making a new underlay to appear on the ground beneath the fire. There's several variants of the fire (flipped horizontally, or some fire icon animation states slower) that could make it a bit disjoined possibly if those aren't removed

**Comment by Azrun** [comment]
> is it feasible to add some overlay or sprite change to hide/obfuscate the really visible tiling? that's my one holdup here versus stuff like hotspots.

This 100%. Using some form of connectedness and having it look like a big fire instead of a bunch of small big fires would go a long way.  This especially stands out as blob is delivered which essentially does something similarly.

**Comment by Tarmunora** [comment]
Haven't tested this rigorously at all, but the chemfires seem to put off a lot less light than gas fires (at least for like, standard red fire)

**Comment by FlameArrow57** [comment]
> Haven't tested this rigorously at all, but the chemfires seem to put off a lot less light than gas fires (at least for like, standard red fire)

Fixed (see below for reference)

image

image

image

**Comment by Sovexe** [comment]
Appears to have introduced a (sometimes server crashing) runtime

```dm
File code/modules/chemistry/Chemistry-Reagents.dm 
Line 165 
Error Cannot read null.aquaphobic 
Usr The monkey () 

 Description
proc name: reaction mob (/datum/reagent/proc/reaction_mob)
  source file: code/modules/chemistry/Chemistry-Reagents.dm,165
  usr: the monkey (/mob/living/carbon/human/npc/monkey)
  src: blood (/datum/reagent/blood)
  usr.loc: the floor (193,158,1) (/turf/simulated/floor)
  call stack:
blood (/datum/reagent/blood): reaction mob(the monkey (/mob/living/carbon/human/npc/monkey), 1, 7.5, 0, 7.5)
blood (/datum/reagent/blood): reaction mob(the monkey (/mob/living/carbon/human/npc/monkey), 1, 7.5, 0, 7.5)
blood (/datum/reagent/blood): reaction mob chemprot layer(the monkey (/mob/living/carbon/human/npc/monkey), 1, 7.5, 0)
/datum/reagents/fluid_group (/datum/reagents/fluid_group): reaction(the monkey (/mob/living/carbon/human/npc/monkey), 1, 7.5, 1, 4, 0, 0)
the monkey (/mob/living/carbon/human/npc/monkey): EnteredFluid(the fluid (/obj/fluid), the floor (193,158,1) (/turf/simulated/floor), 1)
the monkey (/mob/living/carbon/human/npc/monkey): EnteredFluid(the fluid (/obj/fluid), the floor (193,158,1) (/turf/simulated/floor))
the fluid (/obj/fluid): Crossed(the monkey (/mob/living/carbon/human/npc/monkey))
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor), null, null, 0)
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor))
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor))
the monkey (/mob/living/carbon/human/npc/monkey): New(the floor (193,158,1) (/turf/simulated/floor))
the white hole (/obj/whitehole): generate thing("artlab")
the white hole (/obj/whitehole): spew out stuff("artlab")
the white hole (/obj/whitehole): process()
```

**Comment by TobleroneSwordfish** [comment]
Having seen these on live for a bit I really don't like the way the sprites look, they tile really badly and kind of just look like cardboard cutout flame effects. The idea of having separate chem/atmos fire sprites is good, but these don't quite work I think.

**Review by Tarmunora — CHANGES_REQUESTED** [review]
General consensus in dev channels is that the current sprite just doesn't quite work out - something better tiled (and perhaps a bit less busy) will be needed
Otherwise, I think this is pretty much good to go

**Comment by FlameArrow57** [comment]
> General consensus in dev channels is that the current sprite just doesn't quite work out - something better tiled (and perhaps a bit less busy) will be needed Otherwise, I think this is pretty much good to go

Makes sense! Ultimately it will be up to @Flaborized, or anyone else, if they want to sprite something more suited for this. Personally, I just do not have the spriting ability needed for this, and at least for connected sprites across tiles, a number more icon states would need to be sprited (15 at least, using blob tiles per reference, plus possibly animated states). Once sprited though, I'm willing to make any code needed to support

**Comment by Flaborized** [comment]
> > General consensus in dev channels is that the current sprite just doesn't quite work out - something better tiled (and perhaps a bit less busy) will be needed Otherwise, I think this is pretty much good to go
> 
> Makes sense! Ultimately it will be up to @Flaborized, or anyone else, if they want to sprite something more suited for this. Personally, I just do not have the spriting ability needed for this, and at least for connected sprites across tiles, a number more icon states would need to be sprited (15 at least, using blob tiles per reference, plus possibly animated states). Once sprited though, I'm willing to make any code needed to support

I also said this in the devchat discussion about the sprites but I don't have enough time/energy to draw up that many fire sprites :sheltersweat: . If someone else wants to draw all that then that's fine, or if not then if other devs would in general be ok just using the old sprites that'd also be fine by me (I'm not married to this particular sprite or anything), but I can't personally do that sorry!

**Comment by Azrun** [comment]
https://github.com/goonstation/goonstation/assets/1017374/b6a817aa-b8f6-4c6a-95f4-8fc2fd8410d5

What if something like this? Flames merged horizontally, additional "layer" added for vertical neighbor.  Alpha mask to improve visibility of on-turf items.

**Comment by Azrun** [comment]
image

**Comment by FlameArrow57** [comment]
> ChemFire_Demo.mp4 
> What if something like this? Flames merged horizontally, additional "layer" added for vertical neighbor. Alpha mask to improve visibility of on-turf items.

Looks really nice!

File: code/procs/fireflash.dm
Comment (Azrun) [inline code/procs/fireflash.dm:96]: > Summary
> code/procs/fireflash.dm, line 96: list index out of bounds
> Description
> 
> proc name: fireflash melting (/proc/fireflash_melting)
>   source file: code/procs/fireflash.dm,96

Comment (FlameArrow57) [reply code/procs/fireflash.dm:96]: fixed (below commit)

**Comment by TobleroneSwordfish** [comment]
The changes make the fire look a lot better, I really like this now ❤️

**Pull request merged**
